### PR TITLE
Check that all parts of a pattern match, even if a client only wants information about the 'first match'.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Getting just the 'first match' is an optimization to avoid needing to get allocate
an immutable array when doing pattern matching.  It was not intended to be a way
for pattern matching to avoid ensuring that the entire pattern actually matched.

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
**Customer scenario**

user is typing a #r (i.e. file-system path) directive in the interactive window.  When they type reasonable things like "system.refl" it ends up *not* matching "system.reflection" but does erroneously match against things like "runtime.native.system".  This is a pretty poor completion experience for interactive.

**Bugs this fixes:**

N/A

**Workarounds, if any**

None.

**Risk**

Low.

**Performance impact**

Low.

**Is this a regression from a previous update?**

No.

**Root cause analysis:**

Interactive wasn't getting a lot of attention.  IDE rewrote completoin a few times.  This wasn't a scenario that was really considered or given a lot of attention.  Logic we put in pattern matching for some other scenarios was not valid for interactive.

**How was the bug found?**

@tmat Started working on making this scenario good and discovered this poor experience.

**Comment by CyrusNajmabadi** [comment]
Tagging @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
Tagging @ManishJayaswal

**Comment by ManishJayaswal** [comment]
Approved pending CR and successfully running tests.

**Comment by CyrusNajmabadi** [comment]
Ping.  Need PR @dotnet/roslyn-ide

**Comment by CyrusNajmabadi** [comment]
retest this please

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs [hunk 1]
```diff
@@ -239,33 +239,46 @@ public PatternMatches GetMatches(string candidate, string dottedContainer)
...
         {
-            return MatchPatternChunk(candidate, includeMatchSpans: true,
-                patternChunk: _fullPatternSegment.TotalTextChunk, punctuationStripped: false,
-                fuzzyMatch: false);
+            return MatchPatternSegment(
+                candidate, includeMatchSpans, _fullPatternSegment, 
+                wantAllMatches: false, allMatches: out _, fuzzyMatch: fuzzyMatch);
+        }
+
+        private StringBreaks GetWordSpans(string word)
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs]: :memo: (to self) This addition is just a relocated method

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs [hunk 2]
```diff
@@ -436,12 +449,10 @@ private static bool ContainsSpaceOrAsterisk(string text)
 
             var singleMatch = MatchPatternSegment(candidate, includeMatchSpans, patternSegment,
                 wantAllMatches: true, fuzzyMatch: fuzzyMatch, allMatches: out var matches);
-            if (singleMatch.HasValue)
-            {
-                return ImmutableArray.Create(singleMatch.Value);
-            }
 
-            return matches;
+            return singleMatch.HasValue
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs]: :memo: (to self) this change is a neutral refactoring

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs [hunk 3]
```diff
@@ -239,33 +239,46 @@ public PatternMatches GetMatches(string candidate, string dottedContainer)
...
+            string candidate, bool includeMatchSpans)
         {
             if (SkipMatch(candidate))
             {
                 return null;
             }
 
-            return MatchPatternSegment(candidate, inludeMatchSpans, _fullPatternSegment, wantAllMatches: false, allMatches: out _, fuzzyMatch: false) ??
-                   MatchPatternSegment(candidate, inludeMatchSpans, _fullPatternSegment, wantAllMatches: false, allMatches: out _, fuzzyMatch: true);
+            return GetFirstMatchWorker(candidate, includeMatchSpans, fuzzyMatch: false) ??
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs]: :memo: (to self) The behavior of this method is unchanged

File: src/Features/Core/Portable/Completion/CompletionHelper.cs [hunk 4]
```diff
@@ -112,7 +112,7 @@ public bool MatchesPattern(string text, string pattern, CultureInfo culture)
             if (!culture.Equals(EnUSCultureInfo))
             {
                 patternMatcher = this.GetPatternMatcher(pattern, EnUSCultureInfo);
-                match = patternMatcher.GetFirstMatch(completionItemText);
+                match = patternMatcher.GetFirstMatch(completionItemText, includeMatchSpans);
```
Comment (sharwell) [inline src/Features/Core/Portable/Completion/CompletionHelper.cs]: ❗️ This change is a different bug and needs its own regression test. An acceptable alternative for this pull request would be to explicitly pass `false` (preserve original behavior) and create a separate bug to address the issue later.

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs [hunk 5]
```diff
@@ -239,33 +239,46 @@ public PatternMatches GetMatches(string candidate, string dottedContainer)
...
+                   GetFirstMatchWorker(candidate, includeMatchSpans, fuzzyMatch: true);
         }
 
-        private StringBreaks GetWordSpans(string word)
+        public PatternMatch? GetFirstMatch(
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs]: :bulb: It seems like this method can be removed if we just update `TestGetFirstMatch` to coerce a fuzzy result to null. The current pair of overloads is difficult to understand because instinct tells a developer the one normally used is equivalent to the second except that it provides a default value for `fuzzyMatch`, yet this is not the case in practice.

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs [hunk 6]
```diff
@@ -239,33 +239,46 @@ public PatternMatches GetMatches(string candidate, string dottedContainer)
...
             {
-                return _stringToWordSpans.GetOrAdd(word, _breakIntoWordSpans);
+                return null;
             }
+
+            return GetFirstMatchWorker(candidate, includeMatchSpans, fuzzyMatch);
         }
 
-        internal PatternMatch? MatchSingleWordPattern_ForTestingOnly(string candidate)
+        private PatternMatch? GetFirstMatchWorker(
```
Comment (sharwell) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs]: :bulb: If the overloads of `GetFirstMatch` are unified, then this method can likely be inlined without hurting readability.

**Comment by CyrusNajmabadi** [comment]
> Requesting changes as there is an unrelated localization bug fix which does not have an associated bug and no regression test was provided for it.

Sorry... what are you referring to?

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs (hunk 5 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs]: sure.

File: src/Features/Core/Portable/Completion/CompletionHelper.cs (hunk 4 above)
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/Completion/CompletionHelper.cs]: i'm not sure what you're referring to.

File: src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs (hunk 6 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/PatternMatching/PatternMatcher.cs]: I looked at that and didn't like it.  The current pattern makes it clear that all that's changing is the fuzzyMatch flag.  inlining adds a lot of extra complexity that then has to be checked to ensure it's the same.

**Comment by tmat** [comment]
@dotnet/roslyn-ide Ping.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-ide  Please take a look. This is affecting interactive scenarios.

File: src/Features/Core/Portable/Completion/CompletionHelper.cs (hunk 4 above)
Comment (sharwell) [inline src/Features/Core/Portable/Completion/CompletionHelper.cs]: @CyrusNajmabadi When you removed the default value from the parameter in `GetFirstMatch`, it should have been inlined here with its previous default value, `false`. You instead passed through a different value. If the new value is actually the correct value, that correction is unrelated to the rest of this pull request and needs its own separate bug/test.

File: src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs [hunk 7]
```diff
@@ -4066,7 +4066,7 @@ public class Test
             // TODO...
         }
 
-        [Fact]
+        [Fact(Skip = "https://github.com/dotnet/roslyn/issues/18800")]
```
Comment (dpoeschl) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs]: Should really be a separate PR

Comment (sharwell) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs]: ❗️ Since #18800 is no longer relevant, can you revert this change?

Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Test/Symbol/Symbols/SymbolErrorTests.cs]: sure.

**Comment by CyrusNajmabadi** [comment]
Have added tests to validate fallback behavior works.   Please rereview when you have a chance.
